File: DSA/SAMAGRA_590014891/Day_89_Question_3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int V;
int graph[100][100], rev[100][100];
int visited[100], stack[100], top;

void dfs1(int v) {
    visited[v] = 1;
    for (int i = 0; i < V; i++)
        if (graph[v][i] && !visited[i]) dfs1(i);
    stack[top++] = v;
}

void dfs2(int v) {
    visited[v] = 1;
    for (int i = 0; i < V; i++)
        if (rev[v][i] && !visited[i]) dfs2(i);
}

int kosaraju() {
    memset(visited,0,sizeof(visited));
    top = 0;
    for (int i = 0; i < V; i++)
        if (!visited[i]) dfs1(i);
    memset(visited,0,sizeof(visited));
    int count = 0;
    while (top) {
        int v = stack[--top];
        if (!visited[v]) {
            dfs2(v);
            count++;
        }
    }
    return count;
}

int main() {
    V = 5;
    int edges[][2] = {{1,0},{0,2},{2,1},{0,3},{3,4}};
    int E = 5;
    for (int i=0;i<E;i++) {
        graph[edges[i][0]][edges[i][1]] = 1;
        rev[edges[i][1]][edges[i][0]] = 1;
    }
    printf("%d\n", kosaraju());
    return 0;
}

File: DSA/VenikaPanwar_590013755/Day_89_Question_3.java
public class Day_89_Question_3 {

    // Kosaraju's algorithm for SCC count
    public static int kosaraju(int vertices, int[][] edges) {
        java.util.List<java.util.List<Integer>> graph = new java.util.ArrayList<>();
        java.util.List<java.util.List<Integer>> revGraph = new java.util.ArrayList<>();
        for (int i = 0; i < vertices; i++) {
            graph.add(new java.util.ArrayList<>());
            revGraph.add(new java.util.ArrayList<>());
        }
        for (int[] e : edges) {
            graph.get(e[0]).add(e[1]);
            revGraph.get(e[1]).add(e[0]);
        }
        boolean[] visited = new boolean[vertices];
        java.util.Stack<Integer> stack = new java.util.Stack<>();
        for (int i = 0; i < vertices; i++)
            if (!visited[i]) dfs1(graph, visited, stack, i);

        java.util.Arrays.fill(visited, false);
        int count = 0;
        while (!stack.isEmpty()) {
            int node = stack.pop();
            if (!visited[node]) {
                dfs2(revGraph, visited, node);
                count++;
            }
        }
        return count;
    }

    private static void dfs1(java.util.List<java.util.List<Integer>> graph, boolean[] visited, java.util.Stack<Integer> stack, int node) {
        visited[node] = true;
        for (int nei : graph.get(node))
            if (!visited[nei]) dfs1(graph, visited, stack, nei);
        stack.push(node);
    }

    private static void dfs2(java.util.List<java.util.List<Integer>> graph, boolean[] visited, int node) {
        visited[node] = true;
        for (int nei : graph.get(node))
            if (!visited[nei]) dfs2(graph, visited, nei);
    }

    public static void main(String[] args) {
        int v1 = 5;
        int[][] e1 = {{1,0},{0,2},{2,1},{0,3},{3,4}};
        System.out.println(kosaraju(v1, e1)); // Output: 3

        int v2 = 4;
        int[][] e2 = {{0,1},{1,2},{2,3}};
        System.out.println(kosaraju(v2, e2)); // Output: 4
    }
}
